src/TritonCTS/src/SinkAgglDriver.cpp

```diff
@@ -0,0 +1,446 @@
+//
+// BSD 3-Clause License
+//
+// Copyright (c) 2019, University of California, San Diego.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice, this
+//   list of conditions and the following disclaimer.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the following disclaimer in the documentation
+//   and/or other materials provided with the distribution.
+//
+// * Neither the name of the copyright holder nor the names of its
+//   contributors may be used to endorse or promote products derived from
+//   this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+///////////////////////////////////////////////////////////////////////////////
+
+
+#include <vector>
+#include <algorithm>
+#include <iostream>
+#include <cmath> // for std::pow, std::sqrt
+#include <math.h> // for std::pow, std::sqrt
+#include <cstddef> // for std::ptrdiff_t
+#include <limits> // for std::numeric_limits<...>::infinity()
+#include <algorithm> // for std::fill_n
+#include <stdexcept> // for std::runtime_error
+#include <string> // for std::string
+#include <array> // for std::array
+
+#include <cfloat> // also for DBL_MAX, DBL_MIN
+
+#include "SinkAgglDriver.h"
+
+namespace cts {
+
+//#include "SinkAgglEngine.cpp"
+//
+// Assigns cluster labels (0, ..., nclust-1) to the n points such
+// that the cluster result is split into nclust clusters.
+//
+// Input arguments:
+//   n      = number of observables
+//   merge  = clustering result in R format
+//   nclust = number of clusters
+// Output arguments:
+//   labels = allocated integer array of size n for result
+//
+void SinkAgglClustering::cutreeK(int n, const int* merge, int nclust, int* labels) {
+
+  int k,m1,m2,j,l;
+
+  if (nclust > n || nclust < 2) {
+    for (j=0; j<n; j++) labels[j] = 0;
+    return;
+  }
+
+  // assign to each observable the number of its last merge step
+  // beware: indices of observables in merge start at 1 (R convention)
```

Comment:
Agree with Matt here. This makes the code very hard to maintain for others outside the initial authorship group. We should strive for maximum readability rather than minimum code size.